src/wp-includes/global-styles-and-settings.php

```diff
@@ -255,3 +255,70 @@ function ( $item ) {
 		}
 	}
 }
+
+/**
+ * Whether a theme or its parent have a theme.json file.
+ *
+ * @since 6.2.0
+ *
+ * @return bool
+ */
+function wp_theme_has_theme_json() {
+	/*
+	 * By using the 'theme_json' group, this data is marked to be non-persistent across requests.
+	 * See `wp_cache_add_non_persistent_groups` in src/wp-includes/load.php and other places.
+	 *
+	 * The rationale for this is to make sure derived data from theme.json
+	 * is always fresh from the potential modifications done via hooks
+	 * that can use dynamic data (modify the stylesheet depending on some option,
+	 * settings depending on user permissions, etc.).
+	 * See some of the existing hooks to modify theme.json behaviour:
+	 * https://make.wordpress.org/core/2022/10/10/filters-for-theme-json-data/
+	 *
+	 * A different alternative considered was to invalidate the cache upon certain
+	 * events such as options add/update/delete, user meta, etc.
+	 * It was judged not enough, hence this approach.
+	 * See https://github.com/WordPress/gutenberg/pull/45372
+	 */
+	$cache_group       = 'theme_json';
+	$cache_key         = 'wp_theme_has_theme_json';
+	$theme_has_support = wp_cache_get( $cache_key, $cache_group );
+
+	/*
+	 * $theme_has_support is stored as an int in the cache.
+	 *
+	 * The reason not to store it as a boolean is to avoid working
+	 * with the $found parameter which apparently had some issues in some implementations
+	 * https://developer.wordpress.org/reference/functions/wp_cache_get/
+	 *
+	 * Ignore cache when `WP_DEBUG` is enabled, so it doesn't interfere with the theme developers workflow.
+	 */
+	if ( ! WP_DEBUG && is_int( $theme_has_support ) ) {
+		return (bool) $theme_has_support;
+	}
+
+	// Does the theme have its own theme.json?
+	$theme_has_support = is_readable( get_stylesheet_directory() . '/theme.json' );
+
+	// Look up the parent if the child does not have a theme.json.
+	if ( ! $theme_has_support ) {
+		$theme_has_support = is_readable( get_template_directory() . '/theme.json' );
+	}
```

Comment:
Consider https://github.com/WordPress/gutenberg/pull/45831

Reply:
That could be a future change, indeed. We need to measure the performance impact as well. If/when we do that, the corresponding backport should be filled. How is that a blocker for this PR?

Reply:
I am saying, we are have months to go until the 6.2 release, there is no need to merge this change until ALL the PRs on the gutenberg are resolved. I am in no hurry to merge this, let's get it right instead of hurshing changes into core.

Reply:
> we are have months to go until the 6.2 release, there is no need to merge this change until...
> ...
> I am in no hurry to merge this...

I'm actually trying to do exactly the opposite, the "merge early, merge often" strategy. :) 

We are merging to WP trunk (6.2-alpha) so no point in delaying it. On the opposite, if there are any bugs the chances to discover them increase if they are also in core trunk, not just released as plugin. Then if something is fixed/patched "upstream" in the plugin, it can be added to trunk too, no pressure.

Merging early and having more chances to discover bugs seems a lot better than merging in the last minute, and eventually having to fix/patch bugs during RC :)

Reply:
I didn't say this could not be merged. Just that is should not be backported to 6.1.2.